src/osd/OSDMap.cc

```diff
@@ -1770,9 +1770,12 @@ uint64_t OSDMap::get_features(int entity_type, uint64_t *pmask) const
     features |= CEPH_FEATURE_CRUSH_MSR;
   mask |= CEPH_FEATURES_CRUSH;
 
-  if (!pg_upmap.empty() || !pg_upmap_items.empty() || !pg_upmap_primaries.empty())
+  if (!pg_upmap.empty() || !pg_upmap_items.empty())
     features |= CEPH_FEATUREMASK_OSDMAP_PG_UPMAP;
   mask |= CEPH_FEATUREMASK_OSDMAP_PG_UPMAP;
+  if (!pg_upmap_primaries.empty())
+    features |= CEPH_FEATUREMASK_SERVER_REEF;
+  mask |= CEPH_FEATUREMASK_SERVER_REEF;
```

Comment:
Did you test running `ceph osd set-require-min-compat-client reef` with an older client connected to the cluster?  I think `ceph_release_features()` in `src/common/ceph_strings.cc` should be updated as well for it to ever fail with `cannot set require_min_compat_client to reef: ... missing ... add --yes-i-really-mean-it to do it anyway`.

Reply:
All tries with `ceph -w` from `quincy`:

```
[rzarzynski@o06 build]$ bin/ceph daemon mon.a sessions | jq  -jr '.[] | .name, "\t", .con_features, "\t", .con_features_hex, "\n"' | grep client
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
client.?        4540701547738038271     3f03cffffffdffff
client.?        4540138320759226367     3f01cfbf7ffdffff
```

### `reef`
```
[rzarzynski@o06 build]$ bin/ceph osd set-require-min-compat-client reef
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2024-05-29T21:25:42.210+0000 7ff127697700 -1 WARNING: all dangerous and experimental features are enabled.
2024-05-29T21:25:42.218+0000 7ff127697700 -1 WARNING: all dangerous and experimental features are enabled.
set require_min_compat_client to reef
```

### `squid`
```
[rzarzynski@o06 build]$ bin/ceph osd set-require-min-compat-client squid
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2024-05-29T22:39:08.411+0000 7faf316a9700 -1 WARNING: all dangerous and experimental features are enabled.                                                                                                        
2024-05-29T22:39:08.420+0000 7faf316a9700 -1 WARNING: all dangerous and experimental features are enabled.                                                                                                        
set require_min_compat_client to squid
```

This is broader than the `upmap-primary` – squid introduces e.g. MSR.  Actually,  `ceph_release_features()` hasn't been touched since inception in 2017. Looks like a topic for dedicated bug.

Reply:
> Actually, `ceph_release_features()` hasn't been touched since inception in 2017.

Do you recall any client-facing feature bits similar in effect to `OSDMAP_PG_UPMAP` or a new CRUSH tunable added since then?  Among such cluster-wide and breaking features only `CEPHX_V2` comes to mind, but it was handled with a separate knob.  So it might just be upmap-primary and MSR -- both relatively recent.

> Looks like a topic for dedicated bug.

A dedicated ticket makes sense for more investigation and handling the MSR part of it.  However, the upmap-primary part of it must land in 18.2.3 along with this fix.  `ceph osd set-require-min-compat-client reef` not failing in the face of an older client after upmap-primary gets tied to reef for real is a blocker IMO.

Reply:
Huh, looks we haven't been messing with the compatibility for a while.
Created https://tracker.ceph.com/issues/66297 to track the MSR stuff.